Speed up sniffers by using lru_cache on ffprobe, skip binary sniffers on text files

mvdbeek opened the pull request:
For small files the majority of time is being spent calling `ffprobe` on the same file repeatedly (if `ffprobe` is present on the
system). Since the call is always the same this is a prime candidate for using lru_cache.

This saves around 450ms per uploaded file, giving approximately a 2X speedup for small files.
(The total upload time for small files is dominated by imports, interpreter startup and building the datatypes registry, I'm working on
that).

Before:
```
[2022-04-01 12:47:46,755: WARNING/main]          36265 function calls (35608 primitive calls) in 0.785 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:81(_request_to_galaxy_json)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:92(_fetch_target)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:409(elements_tree_map)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:364(_resolve_item_capture_error)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:160(_resolve_item)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:230(_resolve_item_with_primary)
        1    0.000    0.000    0.783    0.783 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/upload_util.py:32(handle_upload)
        1    0.000    0.000    0.782    0.782 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:812(handle_uploaded_dataset_file_internal)
        2    0.000    0.000    0.780    0.390 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:305(guess_ext)
        2    0.005    0.002    0.779    0.390 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:629(run_sniffers_raw)
       14    0.001    0.000    0.723    0.052 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:377(check_output)
       14    0.004    0.000    0.722    0.052 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:464(run)
       14    0.001    0.000    0.653    0.047 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:1090(communicate)
  386/316    0.652    0.002    0.653    0.002 {method 'read' of '_io.BufferedReader' objects}
       10    0.000    0.000    0.644    0.064 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:14(ffprobe)
        2    0.000    0.000    0.178    0.089 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:173(sniff)
        2    0.000    0.000    0.122    0.061 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:227(sniff)
        2    0.000    0.000    0.118    0.059 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:200(sniff)
        2    0.000    0.000    0.116    0.058 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:191(sniff)
        2    0.000    0.000    0.116    0.058 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:209(sniff)
        4    0.000    0.000    0.079    0.020 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/binary.py:2943(sniff)
```

After:
```
[2022-04-01 12:45:34,689: WARNING/main]          32639 function calls (31982 primitive calls) in 0.324 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.324    0.324 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:81(_request_to_galaxy_json)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:92(_fetch_target)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:409(elements_tree_map)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:364(_resolve_item_capture_error)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:160(_resolve_item)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:230(_resolve_item_with_primary)
        1    0.000    0.000    0.322    0.322 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/upload_util.py:32(handle_upload)
        1    0.000    0.000    0.322    0.322 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:812(handle_uploaded_dataset_file_internal)
        2    0.000    0.000    0.319    0.160 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:305(guess_ext)
        2    0.003    0.002    0.319    0.159 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:629(run_sniffers_raw)
        6    0.001    0.000    0.265    0.044 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:464(run)
        6    0.000    0.000    0.231    0.038 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:1090(communicate)
        2    0.000    0.000    0.210    0.105 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:179(sniff)
       10    0.000    0.000    0.209    0.021 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:22(ffprobe)
        2    0.000    0.000    0.209    0.104 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:15(_ffprobe)
        2    0.000    0.000    0.194    0.097 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:1926(_communicate)
        4    0.000    0.000    0.192    0.048 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/selectors.py:403(select)
        4    0.192    0.048    0.192    0.048 {method 'poll' of 'select.poll' objects}
        4    0.000    0.000    0.057    0.014 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/binary.py:2943(sniff)
        4    0.000    0.000    0.057    0.014 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/pytho
```

We should also consider determining if the file is likely binary and then skip the sniff methods for binary datatypes.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mvdbeek committed:
Speed up sniffers by using lru_cache on ffprobe

For small files the majority of time is being spent calling
`ffprobe` on the same file repeatedly (if `ffprobe` is present on the
system). Since the call is always
the same this is a prime candidate for using lru_cache.

This saves around 450ms per uploaded file, giving approximately a 2X
speedup for small files.
(The total upload time for small files is dominated by imports,
interpreter startup and building the datatypes registry, I'm working on
that).

Before:
```
[2022-04-01 12:47:46,755: WARNING/main]          36265 function calls (35608 primitive calls) in 0.785 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:81(_request_to_galaxy_json)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:92(_fetch_target)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:409(elements_tree_map)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:364(_resolve_item_capture_error)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:160(_resolve_item)
        1    0.000    0.000    0.785    0.785 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:230(_resolve_item_with_primary)
        1    0.000    0.000    0.783    0.783 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/upload_util.py:32(handle_upload)
        1    0.000    0.000    0.782    0.782 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:812(handle_uploaded_dataset_file_internal)
        2    0.000    0.000    0.780    0.390 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:305(guess_ext)
        2    0.005    0.002    0.779    0.390 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:629(run_sniffers_raw)
       14    0.001    0.000    0.723    0.052 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:377(check_output)
       14    0.004    0.000    0.722    0.052 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:464(run)
       14    0.001    0.000    0.653    0.047 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:1090(communicate)
  386/316    0.652    0.002    0.653    0.002 {method 'read' of '_io.BufferedReader' objects}
       10    0.000    0.000    0.644    0.064 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:14(ffprobe)
        2    0.000    0.000    0.178    0.089 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:173(sniff)
        2    0.000    0.000    0.122    0.061 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:227(sniff)
        2    0.000    0.000    0.118    0.059 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:200(sniff)
        2    0.000    0.000    0.116    0.058 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:191(sniff)
        2    0.000    0.000    0.116    0.058 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:209(sniff)
        4    0.000    0.000    0.079    0.020 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/binary.py:2943(sniff)
```

After:
```
[2022-04-01 12:45:34,689: WARNING/main]          32639 function calls (31982 primitive calls) in 0.324 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.324    0.324 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:81(_request_to_galaxy_json)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:92(_fetch_target)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:409(elements_tree_map)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:364(_resolve_item_capture_error)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:160(_resolve_item)
        1    0.000    0.000    0.323    0.323 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:230(_resolve_item_with_primary)
        1    0.000    0.000    0.322    0.322 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/upload_util.py:32(handle_upload)
        1    0.000    0.000    0.322    0.322 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:812(handle_uploaded_dataset_file_internal)
        2    0.000    0.000    0.319    0.160 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:305(guess_ext)
        2    0.003    0.002    0.319    0.159 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:629(run_sniffers_raw)
        6    0.001    0.000    0.265    0.044 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:464(run)
        6    0.000    0.000    0.231    0.038 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:1090(communicate)
        2    0.000    0.000    0.210    0.105 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:179(sniff)
       10    0.000    0.000    0.209    0.021 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:22(ffprobe)
        2    0.000    0.000    0.209    0.104 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/media.py:15(_ffprobe)
        2    0.000    0.000    0.194    0.097 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:1926(_communicate)
        4    0.000    0.000    0.192    0.048 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/selectors.py:403(select)
        4    0.192    0.048    0.192    0.048 {method 'poll' of 'select.poll' objects}
        4    0.000    0.000    0.057    0.014 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/binary.py:2943(sniff)
        4    0.000    0.000    0.057    0.014 /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/pytho
```

We should also consider determining if the file is likely binary and
then skip the sniff methods for binary datatypes.

mvdbeek committed:
Trust check_binary heuristic

That brings a single run of the sniffer on a simple text file down to ~15 ms:
```
[2022-04-01 13:11:27,658: WARNING/main]          7441 function calls (7412 primitive calls) in 0.013 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.013    0.013 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:81(_request_to_galaxy_json)
        1    0.000    0.000    0.013    0.013 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:92(_fetch_target)
        1    0.000    0.000    0.013    0.013 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:409(elements_tree_map)
        1    0.000    0.000    0.013    0.013 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:364(_resolve_item_capture_error)
        1    0.000    0.000    0.013    0.013 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:160(_resolve_item)
        1    0.000    0.000    0.013    0.013 /Users/mvandenb/src/metadata_embed/lib/galaxy/tools/data_fetch.py:230(_resolve_item_with_primary)
        1    0.000    0.000    0.012    0.012 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/upload_util.py:32(handle_upload)
        1    0.000    0.000    0.012    0.012 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:812(handle_uploaded_dataset_file_internal)
        2    0.000    0.000    0.010    0.005 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:305(guess_ext)
        2    0.001    0.001    0.008    0.004 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:629(run_sniffers_raw)
        8    0.000    0.000    0.003    0.000 /Users/mvandenb/src/metadata_embed/lib/galaxy/util/compression_utils.py:95(get_fileobj_raw)
      368    0.000    0.000    0.003    0.000 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:215(iter_headers)
       57    0.002    0.000    0.002    0.000 {built-in method io.open}
       50    0.000    0.000    0.002    0.000 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:236(get_headers)
       11    0.000    0.000    0.002    0.000 /Users/mvandenb/src/metadata_embed/lib/galaxy/util/checkers.py:89(check_gzip)
       10    0.000    0.000    0.002    0.000 /Users/mvandenb/src/metadata_embed/lib/galaxy/util/checkers.py:164(is_gzip)
        2    0.000    0.000    0.002    0.001 /Users/mvandenb/src/metadata_embed/lib/galaxy/datatypes/sniff.py:542(__init__)
       30    0.001    0.000    0.001    0.000 {method 'read' of '_io.BufferedReader' objects}
        6    0.000    0.000    0.001    0.000 /Users/mvandenb/src/metadata_embed/lib/galaxy/util/compression_utils.py:56(get_fileobj)
```

nsoranzo commented on lib/galaxy/datatypes/sniff.py:
```diff
@@ -846,11 +846,6 @@ def handle_uploaded_dataset_file_internal(
         guessed_ext = ext
         if ext in AUTO_DETECT_EXTENSIONS:
             guessed_ext = guess_ext(converted_path, sniff_order=datatypes_registry.sniff_order, is_binary=is_binary)
-            guessed_datatype = datatypes_registry.get_datatype_by_extension(guessed_ext)
-            if not is_binary and guessed_datatype.is_binary:
-                # It's possible to have a datatype that is binary but not within the first 1024 bytes,
-                # so check_binary might return a false negative. This is for instance true for PDF files
-                is_binary = True
```
As explained in the comment, we can't really trust `check_binary()` when it returns False, only when it returns True, e.g.:
```python
from galaxy.util.checkers import check_binary

check_binary("lib/galaxy/datatypes/test/454Score.pdf")
```

returns `False`. I'd keep this.

mvdbeek replied:
We should improve check_binary then. There's no reason we need to spend so much time in the sniffers that are not guaranteed to be perfect anyway.

mvdbeek replied:
The comment also predated the improvements in check_binary. We can combine this with a magic check using python-magic (should handle the PDF case), increase the number of bytes to check to what we do with FilePrefix, and we could start by running through the non-binary sniffers first.

nsoranzo commented on lib/galaxy/datatypes/sniff.py:
```diff
@@ -640,6 +640,8 @@ def run_sniffers_raw(filename_or_file_prefix: Union[str, FilePrefix], sniff_orde
         successfully discovered.
         """
+        if is_binary is not None and is_binary != datatype.is_binary:
+            continue
```
```suggestion
        if is_binary and not datatype.is_binary:
            continue
```

mvdbeek committed:
Better binary detection with FilePrefix & libmagic

FilePrefix reads the first Megabytes of a file, which is much more
than the previously used `check_binary` which read only 1KB.

Since we're re-using FilePrefix instance this shouldn't have any impact
on performace. In addition we're now using `python-magic` to detect
the encoding and mime-type. We're using the mime-type detection
for PDF only at this point, but we're also using the detected encoding
if we encounter non-utf8 content.

mvdbeek committed:
Use gzip as string for compressed format

mvdbeek committed:
Fix sniffing for converted datasets

We can't continue with the same FilePrefix here. We could probably be
smarter about this.

mvdbeek committed:
Skip newline / space to tab conversion for keep-compressed files

The binary check in FilePrefix occurs on the already decompressed byte
content, while the previous check ran on the converted path. For
keep-compressed datatypes that meant the previous check would return
false, meaning we'd skip the conversion.

mvdbeek committed:
Fix xlsx sniffing

mvdbeek committed:
Do sniff datatypes for which binary/text distinction is unclear

mvdbeek committed:
Improve is_binary check

mvdbeek committed:
Fix typing for is_binary

davelopez commented on lib/galaxy/datatypes/data.py:
```diff
@@ -10,9 +10,11 @@
     Any,
     Dict,
     List,
+    Literal,
```
I think you need to import `Literal` from `typing_extensions`

mvdbeek committed:
Better way to deal with auto_decompress

jmchilton commented on lib/galaxy/tools/data_fetch.py:
```diff
@@ -496,7 +496,7 @@ def __init__(self, request, registry, working_directory, allow_failed_collection
         self.check_content = request.get("check_content", True)
         self.to_posix_lines = request.get("to_posix_lines", False)
         self.space_to_tab = request.get("space_to_tab", False)
-        self.auto_decompress = request.get("auto_decompress", False)
```
Does this change what the API does by default? It looks like I wanted to make callers indicate this if they want the data to be mucked with - but I might have not done that right.

mvdbeek replied:
It does, and I think I found a better way so we can avoid this.

mvdbeek committed:
Set compressed=False on bcf_uncompressed

mvdbeek committed:
Speed up mrcfile sniffer

This is another good speedup for binary file sniffing.

mvdbeek committed:
Fix wiff.tar sniffing

mvdbeek committed:
Fix wiff sniffing by special-casing tar datatypes
